### ISSUE TITLE
[Bug] Incorrect Result using different num_splits

### ISSUE THREAD

**Issue by ghostplant** [Dao-AILab/flash-attention#1647]
```py
import torch
import flash_attn

B, S, HI, HO, M = 1, 4096, 16, 8, 128
q = torch.randn([B, 1, HI, M]).bfloat16().cuda()
k = torch.randn([B, S, HO, M]).bfloat16().cuda()
v = torch.randn([B, S, HO, M]).bfloat16().cuda()

cache_seqlens = torch.tensor([S], dtype=torch.int32).cuda()

print(flash_attn.flash_attn_func(q, k, v))

print(flash_attn.flash_attn_with_kvcache(q, k, v, cache_seqlens= cache_seqlens, num_splits=16))

print(flash_attn.flash_attn_with_kvcache(q, k, v, cache_seqlens= cache_seqlens, num_splits=32))
```

Outputs:
```txt
>>> print(flash_attn.flash_attn_func(q, k, v))
tensor([[[[-0.0508,  0.0334, -0.0354,  ...,  0.0076, -0.0121,  0.0095],
          [-0.0054,  0.0189, -0.0148,  ...,  0.0079, -0.0142,  0.0242],
          [-0.0732,  0.0608,  0.0155,  ..., -0.0284,  0.0167, -0.0156],
          ...,
          [-0.0408,  0.0171, -0.0530,  ...,  0.0079,  0.0354, -0.0114],
          [ 0.0400,  0.0422, -0.0203,  ..., -0.0033, -0.0044, -0.0080],
          [ 0.0408,  0.0369,  0.0109,  ...,  0.0325,  0.0171,  0.0366]]]],
       device='cuda:0', dtype=torch.bfloat16)
>>> 
>>> print(flash_attn.flash_attn_with_kvcache(q, k, v, cache_seqlens= cache_seqlens, num_splits=16))
tensor([[[[-0.0505,  0.0334, -0.0352,  ...,  0.0076, -0.0121,  0.0096],
          [-0.0018,  0.0359,  0.0070,  ...,  0.0017, -0.0114,  0.0233],
          [-0.0728,  0.0608,  0.0154,  ..., -0.0300, -0.0013, -0.0183],
          ...,
          [-0.0403, -0.0004, -0.0576,  ..., -0.0264,  0.0135, -0.0031],
          [ 0.0374,  0.0693, -0.0203,  ..., -0.0181, -0.0043, -0.0229],
          [ 0.0311,  0.0128,  0.0053,  ...,  0.0118,  0.0311,  0.0884]]]],
       device='cuda:0', dtype=torch.bfloat16)
>>> 
>>> print(flash_attn.flash_attn_with_kvcache(q, k, v, cache_seqlens= cache_seqlens, num_splits=32))
tensor([[[[-0.0442,  0.0255, -0.0310,  ...,  0.0047, -0.0079,  0.0012],
          [ 0.0114,  0.0306, -0.0254,  ...,  0.0247,  0.0039,  0.0223],
          [-0.0786,  0.0864,  0.0042,  ..., -0.0177,  0.0105, -0.0176],
          ...,
          [-0.0151,  0.0315, -0.0767,  ..., -0.2285,  0.1089,  0.0260],
          [ 0.0376,  0.0481, -0.0322,  ..., -0.0070, -0.0038, -0.0022],
          [ 0.0569,  0.0271,  0.0211,  ..., -0.0190,  0.0325, -0.0303]]]],
       device='cuda:0', dtype=torch.bfloat16)
```

**Comment by ghostplant** [comment]
The values between `num_splits=16` and `num_splits=32` are huge.

**Comment by tridao** [comment]
You should compare the difference with a "gold" answer, e.g, by computing the attention in the standard way with Q, K, V in fp32 or fp64.

**Comment by ghostplant** [comment]
Sure, but I think it doesn't explain why `num_splits=16` and `num_splits=32` are quite different, which seems not just a precision issue.

**Comment by tridao** [comment]
Please compare to the gold answer

**Comment by ds-kczerski** [comment]
@ghostplant Tridao is right, you should have tests that allow for some numerical error (e.g., due to different operation orderings, float16 operations, etc.). For example, you can write a PyTorch baseline as a gold reference, determine an appropriate tolerance, and then compare against it.
```
torch.allclose(basic_q_grad, test_q_grad, atol=5e-3)
```

**Comment by ghostplant** [comment]
@ds-kczerski OK, based on the comparison, `torch.allclose([0.0884], [-0.0303], atol=5e-3)` get False results, which doesn't pass the accuracy test.

**Comment by tridao** [comment]
Please compare both to the gold result, not both to each other.

**Comment by ds-kczerski** [comment]
`torch.allclose` with `atol=5e-3` is sufficient for my unit tests to verify gradient correctness when using my modified `flash_attn_func` without the KV cache. If I'm not mistaken, K and V are cached in FP16, so I would expect a larger numerical error.

Also, please take a look at Dao’s example above - it’s a perfect demonstration of how even a simple `div + mul` operation can result in a surprisingly large error (in one out of 1,000 floats).

In your case, you should implement a naive PyTorch version with a KV cache mechanism simulated in pythorch in FP16, and then use that as your golden reference for comparison.

**Comment by ds-kczerski** [comment]
One small additional note: in machine learning, high numerical precision is often not critically important. In the case of the attention mechanism, what's important is that independent vectors - such as the gold-standard q1, q2 and their numerically slightly different counterparts q1', q2' - maintain similar angles. This ensures that the attention mechanism behaves similarly.

FP16 vs FP32: Attention mechanisms are generally robust to small perturbations in Q/K/V vectors - as long as relative geometry (e.g., cosine similarity) is preserved, which is often true even with reduced precision like FP16.

Training convergence and final metrics matter more than exact numerical match of intermediate values. In essence, as long as the training converges and the quality metrics are comparable to those of the reference implementation, slight numerical differences are usually acceptable.
